site/scripts/generate-sitemap.ts

```diff
@@ -1,12 +1,19 @@
 import { writeFileSync } from "fs";
 import path from "path";
+import chalk from "chalk";
 import { generateSiteMap } from "../src/lib/sitemap";
 
-const sitemap = generateSiteMap();
+const script = async () => {
+  console.log(chalk.bold("Generating sitemap..."));
 
-writeFileSync(
-  path.join(process.cwd(), `site-map.json`),
-  JSON.stringify(sitemap, null, "\t"),
-);
+  const sitemap = generateSiteMap();
 
-console.log("✅ Generated site map");
+  writeFileSync(
+    path.join(process.cwd(), `site-map.json`),
+    JSON.stringify(sitemap, null, "\t"),
+  );
+
+  console.log("✅ Site map generated");
+};
+
+export default script();
```

Comment:
To help enforce constancy between the logging messages's formatting using `chalk` and use of emojis, what if we create a wrapper function like this:
```TypeScript
const runScript = async (
  script: () => Promise<void>,
  config: { startMsg: string; successMsg: string; errorMsg: string; }
) => {
  try {
    console.log(chalk.bold(`${config.startMsg}...`));
    await script();
    console.log(`✅ ${config.successMsg}`);
  } catch (error) {
    console.log(`❌ ${config.errorMsg}: `, error)
  }
}
```
(this could also be simplified to just accepting a `scriptName: string` argument, and then logging `Running script ${scriptName}...`, `✅ Completed script "${scriptName}"`, etc)

This would let developers focus on writing the functionality of the scripts:

```TypeScript
...
const script = async () => {
  const sitemap = generateSiteMap();

  writeFileSync(
    path.join(process.cwd(), `site-map.json`),
    JSON.stringify(sitemap, null, "\t"),
  );
};

export default runScript(script, {
  startMsg: "Generating sitemap",
  successMsg: "Site map generated",
  errorMsg: "Could not generate sitemap"
});
```

Reply:
Thanks Ben! I tried creating similar abstractions for scripts in some side projects, but found them more costly than beneficial. Imagine we have a script that generates a file and let’s say we don’t want to override the result unless `FORCE=true` is provided. So our messages will look something like:

```
[bold] Generating our file...
[magenta] File /x/y/z generated.
```

```
[bold] Generating our file...
[gray] File /x/y/z already exists use FORCE=true to overwrite it.
```

```
[bold] Generating our file (with FORCE=true)...
[magenta] File x/y/z generated.
```

As you can see, `startMsg` and `successMsg` vary. File path (`/x/y/z`) can vary too if it’s in some configurable `DATA_DIR_PATH` directory. This is just one possible scenario of what a script might want to do – the reality will be full of diversity.

My current thinking is that just using standard logging functions plus `chalk` is simple, flexible and solid. There is not much added repetition and the cost of forgetting to log something is quite small. On the other hand, we don’t need to learn how to use a new wrapper abstraction or add features to it.

Here are some example scripts I wrote for a side project: https://github.com/kachkaev/tooling-for-how-old-is-this-house/tree/4e1c420a88fb6c3d75fef17729eeed9c720fadc3/src/scripts. There are tens of them, all ‘wrapper-free’.

The only abstraction I find useful these days is [`ScriptError`](https://github.com/kachkaev/tooling-for-how-old-is-this-house/blob/76bb8ca0d867472757a355695cdfd629b37e84f2/src/shared/helpersForScripts.ts#L182-L193). It is 100% like a normal `Error`, but if a script throws it all the way up, Node does not print the call stack. It’s an alternative to writing this, but with a more familiar syntax.

```ts
console.log(chalk.red("Some friendly message explaining why script failed"));
process.exit(42);

// ↓

throw new ScriptError("Some friendly message explaining why script failed");
```

WDYT?

Reply:
Thanks Alex, that makes sense to me - you're right we shouldn't sacrifice flexibility when abstracting how we log messages in the scripts.

One way to avoid developers from having to worry about usage of `chalk` and emojis could be an abstraction such as `ScriptError` as you suggest, but I don't feel strongly either-way about this. I'm happy for this to be left as is.